### PR TITLE
Feature/bc 609 finished tasks

### PR DESCRIPTION
**Pull request by odalys-dataport**
# Short Description

## Checklist before merging

- [x] QA: In addition to review, the code has been manually tested (if manual testing is possible)
- [x] PO: Any deviation from requirements was agreed with Product-Owner / ticket author / support-team

> Notice: Please keep this Pull-Request as a Draft (or add WIP label), until it is ready to be reviewed

### CODE REVIEW

File: src/components/templates/TasksDashboardMain.vue [hunk 1]
```diff
@@ -175,6 +202,45 @@ export default {
...
+		tabThreeHeader: function () {
+			const tabThree = {
+				icon: "$taskFinished",
+				title: this.$t("components.organisms.TasksDashboardMain.tab.finished"),
+			};
+
+			return tabThree;
+		},
+		emptyState: function () {
+			// TODO - refactor?
```
Comment (CeEv) [inline src/components/templates/TasksDashboardMain.vue:214]: I am not sure, for the moment i have no better idea. I think if we split different views in this component it is cleanup by this process.

File: src/components/templates/TasksDashboardMain.vue [hunk 2]
```diff
@@ -208,6 +274,13 @@ export default {
 				}
 			}
 		},
+		onOpenFinishedTasksTab() {
+			// this only properly works, because we switch between clients when archiving a task and therefor trigger a full reload
+			// we should probably find a better solution :D
+			if (!this.hasFinishedTasks) {
+				FinishedTaskModule.fetchTasks();
```
Comment (CeEv) [inline src/components/templates/TasksDashboardMain.vue:281]: I agree sound it not work like expected.

File: src/serverApi/v3/api.ts [hunk 3]
```diff
@@ -1614,6 +1868,16 @@ export const TaskApiFactory = function (configuration?: Configuration, basePath?
...
             return localVarFp.taskControllerFindAll(skip, limit, options).then((request) => request(axios, basePath));
         },
+        /**
+         * 
+         * @param {number} [skip] Number of elements (not pages) to be skipped
+         * @param {number} [limit] Page limit, defaults to 10.
+         * @param {*} [options] Override http request option.
+         * @throws {RequiredError}
+         */
+        taskControllerFindAllFinished(skip?: number, limit?: number, options?: any): AxiosPromise<FinishedTaskListResponse> {
```
Comment (CeEv) [inline src/serverApi/v3/api.ts:1878]: @uidp Do you know why taskControllerFindAllFinished() is so often added to this file?

File: src/store/finished-tasks.ts [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	status: Status = "";
+
+	_taskApi?: TaskApiInterface;
+
+	@Action
+	async fetchTasks(): Promise<void> {
+		this.resetBusinessError();
+		this.setStatus("pending");
+		try {
+			const response = await this.taskApi.taskControllerFindAllFinished();
```
Comment (CeEv) [inline src/store/finished-tasks.ts:67]: I think you must pass the pagination as parameter. In context of lacy loading it should not replace the context. It should add it to existing. But it make sense to use this as initial loading and a seperate method that add new task for the same query. 
Please ask UX which limit should set for initial laoding and which for each "loadMore" step.

File: src/store/finished-tasks.ts [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+			
+			this.setTasks(response.data.data);
```
Comment (CeEv) [inline src/store/finished-tasks.ts:69]: I do not like this data.data. But it must fix in api client. ^-^

File: src/store/finished-tasks.ts [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+		return this.isReady && this.tasks.length > 0;
+	}
+
+	private get isReady(): boolean {
+		return this.status === "completed";
+	}
+
+	private get taskApi() {
+		if (!this._taskApi) {
+			this._taskApi = TaskApiFactory(
```
Comment (CeEv) [inline src/store/finished-tasks.ts:122]: This is the right one and work for this._taskApit askControllerFindAllFinished?

**Review by CeEv — DISMISSED** [review]

File: locale/es.json [hunk 7]
```diff
@@ -602,14 +603,12 @@
...
 	"pages.tasks.student.subtitleOverDue": "Tareas perdidas",
-	"pages.tasks.student.emptyState.subtitle": "Has completado todas tus tareas. ¡Disfruta de tu tiempo libre!",
-	"pages.tasks.student.emptyState.title": "No tienes ninguna tarea abierta.",
 	"pages.tasks.student.title": "Tareas abiertas",
 	"pages.tasks.teacher.subtitleAssigned": "Tareas abiertas con fecha límite",
 	"pages.tasks.teacher.subtitleNoDue": "Tareas abiertas sin fecha límite",
 	"pages.tasks.teacher.subtitleOverDue": "Tareas expiradas",
-	"pages.tasks.teacher.emptyState.subtitle": "Has calificado todas las tareas. ¡Disfruta de tu tiempo libre!",
-	"pages.tasks.teacher.emptyState.title": "No tienes ninguna tarea abierta.",
+	"pages.tasks.teacher.open.emptyState.subtitle": "Has calificado todas las tareas. ¡Disfruta de tu tiempo libre!",
```
[the remark this answers is not in the document]
Comment (odalys-dataport) [reply locale/es.json:610]: Me too, I started cleaning it up a little in an earlier PR, but when I saw how much differences the spanish file had from the others, I dropped it.

File: src/store/finished-tasks.ts (hunk 4 above)
Comment (odalys-dataport) [reply src/store/finished-tasks.ts:67]: Yes, I can pass limit and skip here. As a first step to have a working code example I just fetched all of them for now. I have reached out to Mona regarding the limit.

File: src/components/templates/TasksDashboardMain.vue [hunk 8]
```diff
@@ -101,33 +122,43 @@ export default {
 		isSubstituteFilterEnabled: () => TaskModule.isSubstituteFilterEnabled,
 		courseFilters: () => TaskModule.getCourseFilters,
 		selectedCourseFilters: () => TaskModule.getSelectedCourseFilters,
+		finishedTasksInitialized: () => FinishedTaskModule.getInitialized,
```
Comment (uidp) [inline src/components/templates/TasksDashboardMain.vue:125]: maybe `isInitialized`? or is it better to generally begin all getter names with `get`?

File: src/components/templates/TasksDashboardMain.vue [hunk 9]
```diff
@@ -190,22 +252,23 @@ export default {
...
-					return this.tasksCountStudent.open[courseName];
-				}
-				if (this.tab === 1) {
-					return this.tasksCountStudent.completed[courseName];
-				}
+			if (this.tab === 0) {
+				return this.isStudent
+					? this.tasksCountStudent.open[courseName]
+					: this.tasksCountTeacher.open[courseName];
 			}
```
Comment (uidp) [inline src/components/templates/TasksDashboardMain.vue:259]: this looks much better than before.

File: src/components/templates/TasksDashboardMain.vue [hunk 10]
```diff
@@ -190,22 +252,23 @@ export default {
...
-					return this.tasksCountTeacher.open[courseName];
-				}
-				if (this.tab === 1) {
-					return this.tasksCountTeacher.drafts[courseName];
-				}
+			if (this.tab === 1) {
+				return this.isStudent
+					? this.tasksCountStudent.completed[courseName]
+					: this.tasksCountTeacher.drafts[courseName];
+			}
```
Comment (uidp) [inline src/components/templates/TasksDashboardMain.vue:264]: this looks much better than before.

File: src/components/templates/TasksDashboardMain.vue [hunk 11]
```diff
@@ -190,22 +252,23 @@ export default {
...
+		},
+		onOpenFinishedTasksTab() {
+			// this only properly works, because we switch between clients when archiving a task and therefor trigger a full reload
+			// we should probably find a better solution :D
+			if (!this.finishedTasksInitialized) {
+				console.log("initial triggered");
```
Comment (uidp) [inline src/components/templates/TasksDashboardMain.vue:270]: should we remove that `console.log` statement?

File: src/components/templates/TasksDashboardStudent.vue [hunk 12]
```diff
@@ -12,26 +12,26 @@
 					:expanded-default="1"
 				>
 					<template v-slot:panelOne>
-						<tasks-list :tasks="noDueDateTasks" type="student" />
+						<tasks-list :tasks="noDueDateTasks" role="student" />
```
Comment (uidp) [inline src/components/templates/TasksDashboardStudent.vue:15]: i like `role` much better than `type`

File: src/store/tasks.unit.ts [hunk 13]
```diff
@@ -11,7 +11,7 @@ type FunctionPropertyNames<T> = {
 
 /**
  * Spy on a TaskFilter method and mock its return value.
- * The mock is valid vor all instances of the TaskFilter class.
+ * The mock is valid for all instances of the TaskFilter class.
```
Comment (uidp) [inline src/store/tasks.unit.ts:14]: thanks :-)

File: src/store/tasks.unit.ts [hunk 14]
```diff
@@ -292,7 +292,7 @@ describe("task store", () => {
 				expect(hasTasks).toBe(true);
 			});
 
-			it("should return false when tasks are emty", () => {
+			it("should return false when tasks are empty", () => {
```
Comment (uidp) [inline src/store/tasks.unit.ts:295]: thanks again :-)

**Review by uidp — APPROVED** [review]

**Review by uidp — APPROVED** [review]

**Review by uidp — APPROVED** [review]

**Pull request merged**